action/protocol/staking/candidate_buckets_indexer.go

```diff
@@ -0,0 +1,119 @@
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+import (
+	"context"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+)
+
+const (
+	// StakingCandidatesNamespace is a namespace to store candidates with epoch start height
+	StakingCandidatesNamespace = "stakingCandidates"
+	// StakingBucketsNamespace is a namespace to store vote buckets with epoch start height
+	StakingBucketsNamespace = "stakingBuckets"
+)
+
+// CandidatesBucketsIndexer is an indexer to store candidates by given height
+type CandidatesBucketsIndexer struct {
+	kvStore db.KVStore
+}
+
+// NewStakingCandidatesBucketsIndexer creates a new StakingCandidatesIndexer
+func NewStakingCandidatesBucketsIndexer(kv db.KVStore) (*CandidatesBucketsIndexer, error) {
+	if kv == nil {
+		return nil, ErrMissingField
+	}
+	return &CandidatesBucketsIndexer{
+		kvStore: kv,
+	}, nil
+}
+
+// Start starts the indexer
+func (cbi *CandidatesBucketsIndexer) Start(ctx context.Context) error {
+	return cbi.kvStore.Start(ctx)
+}
+
+// Stop stops the indexer
+func (cbi *CandidatesBucketsIndexer) Stop(ctx context.Context) error {
+	return cbi.kvStore.Stop(ctx)
+}
+
+// PutCandidates puts candidates into indexer
+func (cbi *CandidatesBucketsIndexer) PutCandidates(height uint64, candidates *iotextypes.CandidateListV2) error {
+	candidatesBytes, err := proto.Marshal(candidates)
+	if err != nil {
+		return err
+	}
+	return cbi.kvStore.Put(StakingCandidatesNamespace, byteutil.Uint64ToBytes(height), candidatesBytes)
+}
+
+// GetCandidates gets candidates from indexer given epoch start height
+func (cbi *CandidatesBucketsIndexer) GetCandidates(height uint64, offset, limit uint32) ([]byte, error) {
+	candidateList := &iotextypes.CandidateListV2{}
+	ret, err := cbi.kvStore.Get(StakingCandidatesNamespace, byteutil.Uint64ToBytes(height))
+	if errors.Cause(err) == db.ErrNotExist {
+		return proto.Marshal(candidateList)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if err := proto.Unmarshal(ret, candidateList); err != nil {
+		return nil, err
+	}
+	length := uint32(len(candidateList.Candidates))
+	if offset >= length {
+		return proto.Marshal(&iotextypes.CandidateListV2{})
+	}
+	end := offset + limit
+	if end > uint32(len(candidateList.Candidates)) {
+		end = uint32(len(candidateList.Candidates))
+	}
+	candidateList.Candidates = candidateList.Candidates[offset:end]
+	return proto.Marshal(candidateList)
+}
+
+// PutBuckets puts vote buckets into indexer
+func (cbi *CandidatesBucketsIndexer) PutBuckets(height uint64, buckets *iotextypes.VoteBucketList) error {
+	bucketsBytes, err := proto.Marshal(buckets)
+	if err != nil {
+		return err
+	}
+	return cbi.kvStore.Put(StakingBucketsNamespace, byteutil.Uint64ToBytes(height), bucketsBytes)
```

Comment:
use Uint64ToBytesBigEndian() so that the height key is sorted
and can use kv.Filter() to access a range of data (like epoch 10~20) in case there's such need